### ISSUE TITLE
make_table().to_pandas() error related to globals

### ISSUE THREAD

**Issue by gokceneraslan** [hail-is/hail#5212]
### What you did:

```python
smt = hl.methods.balding_nichols_model(3, 100, 200)
smt = smt.annotate_cols(s=hl.str(smt.sample_idx)).key_cols_by('s')
smt.make_table().to_pandas()
```
### What went wrong (all error messages here, including the full java stack trace):

```
2019-01-26 19:03:28 Hail: INFO: balding_nichols_model: generating genotypes for 3 populations, 100 samples, and 200 variants...
2019-01-26 19:03:28 Hail: INFO: Coerced sorted dataset
2019-01-26 19:03:29 Hail: INFO: Coerced sorted dataset
---------------------------------------------------------------------------
FatalError                                Traceback (most recent call last)
<ipython-input-59-9addf4eaf59b> in <module>
      1 smt = hl.methods.balding_nichols_model(3, 100, 200)
      2 smt = smt.annotate_cols(s=hl.str(smt.sample_idx)).key_cols_by('s')
----> 3 smt.make_table().to_pandas()

<decorator-gen-1366> in to_pandas(self, flatten)

~/anaconda3/lib/python3.6/site-packages/hail/typecheck/check.py in wrapper(__original_func, *args, **kwargs)
    558     def wrapper(__original_func, *args, **kwargs):
    559         args_, kwargs_ = check_all(__original_func, args, kwargs, checkers, is_method=is_method)
--> 560         return __original_func(*args_, **kwargs_)
    561 
    562     return wrapper

~/anaconda3/lib/python3.6/site-packages/hail/table.py in to_pandas(self, flatten)
   2703 
   2704         """
-> 2705         return Env.spark_backend('to_pandas').to_pandas(self, flatten)
   2706 
   2707     @staticmethod

~/anaconda3/lib/python3.6/site-packages/hail/backend/backend.py in to_pandas(self, t, flatten)
     66 
     67     def to_pandas(self, t, flatten):
---> 68         return self.to_spark(t, flatten).toPandas()
     69 
     70     def from_pandas(self, df, key):

~/anaconda3/lib/python3.6/site-packages/hail/backend/backend.py in to_spark(self, t, flatten)
     63         if flatten:
     64             t = t.flatten()
---> 65         return pyspark.sql.DataFrame(t._jt.toDF(Env.hc()._jsql_context), Env.sql_context())
     66 
     67     def to_pandas(self, t, flatten):

~/anaconda3/lib/python3.6/site-packages/py4j/java_gateway.py in __call__(self, *args)
   1255         answer = self.gateway_client.send_command(command)
   1256         return_value = get_return_value(
-> 1257             answer, self.gateway_client, self.target_id, self.name)
   1258 
   1259         for temp_arg in temp_args:

~/anaconda3/lib/python3.6/site-packages/hail/utils/java.py in deco(*args, **kwargs)
    225             raise FatalError('%s\n\nJava stack trace:\n%s\n'
    226                              'Hail version: %s\n'
--> 227                              'Error summary: %s' % (deepest, full, hail.__version__, deepest)) from None
    228         except pyspark.sql.utils.CapturedException as e:
    229             raise FatalError('%s\n\nJava stack trace:\n%s\n'

FatalError: AssertionError: assertion failed: type mismatch:
  name: global
  actual: Struct{bn:Struct{n_populations:Int32,n_samples:Int32,n_variants:Int32,n_partitions:Int32,pop_dist:Array[Int32],fst:Array[Float64],mixture:Boolean},__uid_882:Array[Struct{sample_idx:Int32,pop:Int32,s:String}]}
  expect: Struct{bn:Struct{n_populations:Int32,n_samples:Int32,n_variants:Int32,n_partitions:Int32,pop_dist:Array[Int32],fst:Array[Float64],mixture:Boolean},__cols:Array[Struct{sample_idx:Int32,pop:Int32,s:String}]}

Java stack trace:
is.hail.utils.HailException: Error while typechecking IR:
(MakeStruct
  (bn
    (GetField bn
      (Ref global))))
	at is.hail.utils.ErrorHandling$class.fatal(ErrorHandling.scala:11)
	at is.hail.utils.package$.fatal(package.scala:26)
	at is.hail.expr.ir.TypeCheck$.apply(TypeCheck.scala:16)
	at is.hail.expr.ir.Compile$.apply(Compile.scala:45)
	at is.hail.expr.ir.Compile$.apply(Compile.scala:32)
	at is.hail.expr.ir.Compile$.apply(Compile.scala:77)
	at is.hail.expr.ir.TableMapGlobals$$anonfun$38.apply(TableIR.scala:856)
	at is.hail.expr.ir.TableMapGlobals$$anonfun$38.apply(TableIR.scala:846)
	at is.hail.utils.package$.using(package.scala:587)
	at is.hail.annotations.Region$.scoped(Region.scala:13)
	at is.hail.expr.ir.TableMapGlobals.execute(TableIR.scala:846)
	at is.hail.expr.ir.TableKeyBy.execute(TableIR.scala:237)
	at is.hail.expr.ir.TableMapRows.execute(TableIR.scala:696)
	at is.hail.expr.ir.TableMapGlobals.execute(TableIR.scala:838)
	at is.hail.expr.ir.TableMapRows.execute(TableIR.scala:696)
	at is.hail.expr.ir.Interpret$.apply(Interpret.scala:42)
	at is.hail.table.Table.x$3$lzycompute(Table.scala:211)
	at is.hail.table.Table.x$3(Table.scala:211)
	at is.hail.table.Table.rvd$lzycompute(Table.scala:211)
	at is.hail.table.Table.rvd(Table.scala:211)
	at is.hail.table.Table.toDF(Table.scala:455)
	at sun.reflect.GeneratedMethodAccessor107.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)

java.lang.AssertionError: assertion failed: type mismatch:
  name: global
  actual: Struct{bn:Struct{n_populations:Int32,n_samples:Int32,n_variants:Int32,n_partitions:Int32,pop_dist:Array[Int32],fst:Array[Float64],mixture:Boolean},__uid_882:Array[Struct{sample_idx:Int32,pop:Int32,s:String}]}
  expect: Struct{bn:Struct{n_populations:Int32,n_samples:Int32,n_variants:Int32,n_partitions:Int32,pop_dist:Array[Int32],fst:Array[Float64],mixture:Boolean},__cols:Array[Struct{sample_idx:Int32,pop:Int32,s:String}]}
	at scala.Predef$.assert(Predef.scala:170)
	at is.hail.expr.ir.TypeCheck$._apply(TypeCheck.scala:60)
	at is.hail.expr.ir.TypeCheck$.is$hail$expr$ir$TypeCheck$$check$1(TypeCheck.scala:22)
	at is.hail.expr.ir.TypeCheck$._apply(TypeCheck.scala:237)
	at is.hail.expr.ir.TypeCheck$.is$hail$expr$ir$TypeCheck$$check$1(TypeCheck.scala:22)
	at is.hail.expr.ir.TypeCheck$$anonfun$_apply$18.apply(TypeCheck.scala:213)
	at is.hail.expr.ir.TypeCheck$$anonfun$_apply$18.apply(TypeCheck.scala:213)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at is.hail.expr.ir.TypeCheck$._apply(TypeCheck.scala:213)
	at is.hail.expr.ir.TypeCheck$.apply(TypeCheck.scala:14)
	at is.hail.expr.ir.Compile$.apply(Compile.scala:45)
	at is.hail.expr.ir.Compile$.apply(Compile.scala:32)
	at is.hail.expr.ir.Compile$.apply(Compile.scala:77)
	at is.hail.expr.ir.TableMapGlobals$$anonfun$38.apply(TableIR.scala:856)
	at is.hail.expr.ir.TableMapGlobals$$anonfun$38.apply(TableIR.scala:846)
	at is.hail.utils.package$.using(package.scala:587)
	at is.hail.annotations.Region$.scoped(Region.scala:13)
	at is.hail.expr.ir.TableMapGlobals.execute(TableIR.scala:846)
	at is.hail.expr.ir.TableKeyBy.execute(TableIR.scala:237)
	at is.hail.expr.ir.TableMapRows.execute(TableIR.scala:696)
	at is.hail.expr.ir.TableMapGlobals.execute(TableIR.scala:838)
	at is.hail.expr.ir.TableMapRows.execute(TableIR.scala:696)
	at is.hail.expr.ir.Interpret$.apply(Interpret.scala:42)
	at is.hail.table.Table.x$3$lzycompute(Table.scala:211)
	at is.hail.table.Table.x$3(Table.scala:211)
	at is.hail.table.Table.rvd$lzycompute(Table.scala:211)
	at is.hail.table.Table.rvd(Table.scala:211)
	at is.hail.table.Table.toDF(Table.scala:455)
	at sun.reflect.GeneratedMethodAccessor107.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)




Hail version: 0.2.8-70304a52d33d
Error summary: AssertionError: assertion failed: type mismatch:
  name: global
  actual: Struct{bn:Struct{n_populations:Int32,n_samples:Int32,n_variants:Int32,n_partitions:Int32,pop_dist:Array[Int32],fst:Array[Float64],mixture:Boolean},__uid_882:Array[Struct{sample_idx:Int32,pop:Int32,s:String}]}
  expect: Struct{bn:Struct{n_populations:Int32,n_samples:Int32,n_variants:Int32,n_partitions:Int32,pop_dist:Array[Int32],fst:Array[Float64],mixture:Boolean},__cols:Array[Struct{sample_idx:Int32,pop:Int32,s:String}]}
```

**Comment by gokceneraslan** [comment]
Dropping globals (`smt = smt.drop(smt.bn)`) solves this but no idea how to fix it properly.

**Comment by tpoterba** [comment]
optimizer / compiler bug. Shouldn't be too hard to fix!

**Comment by jigold** [comment]
This bug was fixed already in 83bd98b6f #5132. I did a binary search to find the commit that fixed this.
